### ISSUE TITLE
[python] suppress the warning about categorical feature override

### ISSUE THREAD

**Issue by guolinke** [microsoft/LightGBM#3379]
```
C:\ProgramData\Anaconda3\lib\site-packages\lightgbm\basic.py:1555: UserWarning: categorical_feature in Dataset is overridden.
New categorical_feature is ['C01', 'C02', 'C03', 'C04', 'C05', 'C06', 'C07', 'C08', 'C09', 'C10', 'C11', 'C12', 'C13', 'C14', 'C15', 'C16', 'C17', 'C18', 'C19', 'C20', 'C21', 'C22', 'C23', 'C24', 'C25', 'C26']
  'New categorical_feature is {}'.format(sorted(list(categorical_feature))))
C:\ProgramData\Anaconda3\lib\site-packages\lightgbm\basic.py:1286: UserWarning: Overriding the parameters from Reference Dataset.
  warnings.warn('Overriding the parameters from Reference Dataset.')
C:\ProgramData\Anaconda3\lib\site-packages\lightgbm\basic.py:1098: UserWarning: categorical_column in param dict is overridden.
  warnings.warn('{} in param dict is overridden.'.format(cat_alias))
```
`categorical_column` could be set in both `lgb.train` and `lgb.Dataset` . But this warning seems always show up if setting  `categorical_column`. I think this is quite annoying.

**Comment by onshek** [comment]
I'm willing to take this as my first contribution to the repo.
It seems a new parameter `ignore_warnings` detault to be `True` can be added to related functions, right?

**Comment by XiaozhouWang85** [comment]
From reading other issues, it seems the "correct" way of defining categorical columns is via the Dataset. This might only be a problem when the input is a Pandas dataframe. 

The problem is that when calling `train` the categorical column defaults to auto: 
`categorical_feature (list of strings or int, or 'auto', optional (default="auto"))`

If we had an option of None then that might be a way of getting rid of conflicting settings that require a warning.

**Comment by istavnit** [comment]
Seems to be not a friendly way to suppress this warning:
```
            XX_train = XX_train[keeperCols]
            XX_valid = XX_valid[keeperCols]
            #I have only one categorical feature 'DOW'  
            trainData = Dataset(XX_train,yy_train,feature_name=keeperCols,categorical_feature=['DOW'])
            valid_data = trainData.create_valid(XX_valid,label=yy_valid)
            params["seed"]=theSeed
            bst = lgb.train(params, trainData, valid_sets=[valid_data], categorical_feature=['DOW'],feature_name=keeperCols,
                            num_boost_round=BOOSTROUNDS, early_stopping_rounds=EARLYSTOPROUNDS,verbose_eval=VERBOSEEVAL)

```
results in these warnings:
```
C:\Users\meanc\anaconda3\envs\keras\lib\site-packages\lightgbm\basic.py:1551: UserWarning: Using categorical_feature in Dataset.
  warnings.warn('Using categorical_feature in Dataset.')
C:\Users\meanc\anaconda3\envs\keras\lib\site-packages\lightgbm\basic.py:1555: UserWarning: categorical_feature in Dataset is overridden.
New categorical_feature is ['DOW']
  'New categorical_feature is {}'.format(sorted(list(categorical_feature))))
C:\Users\meanc\anaconda3\envs\keras\lib\site-packages\lightgbm\basic.py:1286: UserWarning: Overriding the parameters from Reference Dataset.
  warnings.warn('Overriding the parameters from Reference Dataset.')
C:\Users\meanc\anaconda3\envs\keras\lib\site-packages\lightgbm\basic.py:1098: UserWarning: categorical_column in param dict is overridden.
  warnings.warn('{} in param dict is overridden.'.format(cat_alias))
```

**Comment by memeplex** [comment]
When I do a grid search using the sklearn api and pass eval_set to fit, I get this warning for every element in the grid (many times!).

I'm just passing a dataframe with categorical features as the train X, never explicitly passing categorical_feature.

I don't think this behavior is desirable.

**Comment by memeplex** [comment]
Indeed it's not necessary to use the sklearn API in order to reproduce the above. I've provided simple instructions in https://github.com/microsoft/LightGBM/issues/3640.

**Comment by tripti0125** [comment]
I get this warning when using scikit-learn wrapper of LightGBM. Dataset passed to LightGBM is through a scikit-learn pipeline which preprocesses the data in a pandas dataframe and produces a numpy array. Note that this input dataset which the model receives is NOT a Pandas dataframe but numpy array. I set the feature_name and categorical_feature parameters in fit() method as this is the only place these can be set, if you're not using LightGBM native Dataset creation.

I think the warning is useful in some situations but superfluous in the case mentioned above.

C:\..\anaconda3\lib\site-packages\lightgbm\basic.py:1286: UserWarning: Overriding the parameters from Reference Dataset.
  warnings.warn('Overriding the parameters from Reference Dataset.')
C:\..\anaconda3\lib\site-packages\lightgbm\basic.py:1098: UserWarning: categorical_column in param dict is overridden.
  warnings.warn('{} in param dict is overridden.'.format(cat_alias))

**Comment by ThomasBourgeois** [comment]
Hi all, 
same issue here : I'm sepcifying categorical features pretty much everywhere : train and val datasets, plus in train, and I still get warnings.. Very strange.
Warnings obtained :
"Overriding the parameters from Reference Dataset.
categorical_column in param dict is overridden."

Code :
`
train_data = lgb.Dataset(train[feats], label=train[target],
                         feature_name=feats,
                         categorical_feature=cat_feats)
val_data = lgb.Dataset(val[feats], label=val[target], reference=train_data,
                          feature_name=feats,
                          categorical_feature=cat_feats)

params = {'objective': 'mean_squared_error', 
        'metric':'rmse', 'eta': 0.1, 'bagging_fraction': 0.5 }
num_round = 300
bst = lgb.train(params, train_data, num_round, valid_sets=[val_data], early_stopping_rounds=20, 
               categorical_feature=cat_feats, feature_name=feats
               )
`

**Comment by memeplex** [comment]
With multiple processes in a grid search it's not event possible to use a context manager to suppress this warning during fit, it seems that the context state is lost somehow, my notebook gets literally flooded of:

```
/Users/carlos/Base/Environments/Jampp/lib/python3.9/site-packages/lightgbm/basic.py:1702: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
/Users/carlos/Base/Environments/Jampp/lib/python3.9/site-packages/lightgbm/basic.py:1702: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
/Users/carlos/Base/Environments/Jampp/lib/python3.9/site-packages/lightgbm/basic.py:1702: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
/Users/carlos/Base/Environments/Jampp/lib/python3.9/site-packages/lightgbm/basic.py:1702: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
/Users/carlos/Base/Environments/Jampp/lib/python3.9/site-packages/lightgbm/basic.py:1702: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
/Users/carlos/Base/Environments/Jampp/lib/python3.9/site-packages/lightgbm/basic.py:1702: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
/Users/carlos/Base/Environments/Jampp/lib/python3.9/site-packages/lightgbm/basic.py:1702: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
....
```

**Comment by memeplex** [comment]
As a workaround I did the following:

```python
    class SilentRegressor(lgb.LGBMRegressor):
        def fit(self, *args, **kwargs):
            with ignore_warnings(UserWarning):
                return super().fit(*args, verbose=False, **kwargs)
```

**Comment by hzy46** [comment]
I did some investigation on this issue. There're mainly two kinds of source for these warnings regarding categorical features.

### If the dataset doens't have a reference, the warnings come from here:

https://github.com/microsoft/LightGBM/blob/7fa07ee2058eaacd41109347afded8c887507eba/python-package/lightgbm/basic.py#L2045-L2074

This function will be called before the Dataset.construct() is called.

One can use the following code to reproduce:

```python
import random
import numpy as np
import pandas as pd
import lightgbm as lgb


Categorical_Feature_When_Construct_Dataset = ["a", "b", "d"]
Categorical_Feature_When_Train = 'auto'


def get_data(N):
    data = []
    labels = []
    for i in range(N):
        sample = {
            "a": random.choice([100, 200, 300, 400]),
            "b": random.choice([222, 333]),
            "c": random.random(),
        }
        if sample["a"] == 200 or sample["a"] == 300:
            if sample["b"] == 333:
                label = 1
            else:
                label = 0
        else:
            label = 0
        labels.append(label)
        data.append(sample)
    features = pd.DataFrame(data)
    features["d"] = pd.Categorical(
        [random.choice(["x", "y", "z"]) for i in range(N)], categories=["x", "y", "z"], ordered=False
    )
    labels = pd.Series(labels)
    return features, labels
 
N = 1000
train_features, train_labels = get_data(N)
test_features, test_labels = get_data(N)
 
lgb_train = lgb.Dataset(train_features, train_labels, categorical_feature=Categorical_Feature_When_Construct_Dataset)


params = {
    'boosting_type': 'gbdt',
    'objective': 'regression',
    'metric': {'l2', 'l1'},
    'num_leaves': 4,
    'learning_rate': 0.5,
    'verbose': 0,
}


gbm = lgb.train(params,
                lgb_train,
                num_boost_round=1,
                categorical_feature=Categorical_Feature_When_Train,
)
 
```

Here, if `Categorical_Feature_When_Construct_Dataset` doesn't equal to `Categorical_Feature_When_Train`, a warning will be reported. These two parameters have the same default value: `auto`. It becomes confusing when thw user only sets `Categorical_Feature_When_Construct_Dataset`, and don't set `Categorical_Feature_When_Train`. His expectation should be using the same categorical feature as the dataset's setting. But an warning is reported in this case because `Categorical_Feature_When_Train` has an default value `auto`, which doesn't align with `Categorical_Feature_When_Construct_Dataset`.

If we use `cf1` to represent `Categorical_Feature_When_Construct_Dataset`, and `cf2` to represent `Categorical_Feature_When_Train`, we can have the following behavior: 

|case| `cf1` |  `cf2` | Current behavior | Warning |
|----| ----- | ----------- | ----|----|
|1| auto | auto | auto | no | 
|2| auto | specific columns | use `cf2` | yes |
|3| specific columns | auto | use `cf1` | yes |
|4| specific columns|  specific columns| use `cf2` | yes if `cf1` and `cf2` are different|

For this first source, my proposal is: ** If the user is using specific columns to override "auto", we don't report the warning. Because the user is just overriding the default parameter. ** It aligns with the current behavior. What we need to do is to remove the warning information for case 2 and case 3 in the table.

### The second source comes from the dataset with a reference:

https://github.com/microsoft/LightGBM/blob/7fa07ee2058eaacd41109347afded8c887507eba/python-package/lightgbm/basic.py#L1778-L1781

This is always reported if the referenced dataset has any categorical features. For the referenced dataset, its `self.params` is changed here:

https://github.com/microsoft/LightGBM/blob/7fa07ee2058eaacd41109347afded8c887507eba/python-package/lightgbm/basic.py#L1498-L1518

For this one, my suggestion is to **ignore categorical features when comparing the params**.

**Comment by shiyu1994** [comment]
Thanks for your detailed analysis. I think the proposed solution is feasible!

**Comment by thisisreallife** [comment]
> As a workaround I did the following:
> 
> ```python
> class SilentRegressor(lgb.LGBMRegressor):
>     def fit(self, *args, **kwargs):
>         with warnings.catch_warnings():
>             warnings.filterwarnings("ignore", category=UserWarning)
>             return super().fit(*args, verbose=False, **kwargs)
> ```
following code is fine, if we do not want to create a new class.
```
import warnings
warnings.filterwarnings("ignore", category=UserWarning)
```

**Comment by Nevermetyou65** [comment]
Is this problem solved??
I still can get this warning even I set auto when construction dataset and train. It's so annoying
